bob-access/src/credentials.rs

```diff
@@ -0,0 +1,149 @@
+use std::net::{IpAddr, SocketAddr};
+use std::marker::PhantomData;
+
+use axum::{
+    async_trait,
+    extract::{FromRequest, RequestParts},
+};
+use tonic::Request;
+
+use crate::{error::Error, extractor::ExtractorExt, Authenticator};
+
+#[derive(Debug, Default, Clone)]
+pub struct Credentials {
+    address: Option<SocketAddr>,
+    kind: Option<CredentialsKind>,
+}
+
+pub struct CredentialsHolder<A: Authenticator> {
+    credentials: Credentials,
+    pd: PhantomData<A>,
+}
+
+impl<A: Authenticator> From<CredentialsHolder<A>> for Credentials {
+    fn from(holder: CredentialsHolder<A>) -> Self {
+        holder.credentials
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum CredentialsKind {
+    Basic { username: String, password: String },
+    Token(String),
+    InterNode(String),
+}
+
+impl CredentialsKind {
+    pub fn is_internode(&self) -> bool {
```

Comment:
Same functions can be added for all `CredentialsKind` values (additionally for `Basic` and for `Token`)